relay-aws-extension/src/aws_extension.rs

```diff
@@ -0,0 +1,193 @@
+/// * sends blocking NextEvent calls to the extensions API to get the next invocation
+///
+/// Each finished invocation immediately polls for the next event. Note that AWS might
+/// freeze the container indefinitely for unused lambdas and this request will also wait
+/// till things are active again.
+pub struct AwsExtension {
+    /// The base url for the AWS Extensions API
+    base_url: String,
+    /// The reqwest client needs to be blocking and with 0 timeout
+    /// because the container might get frozen if lambda is unused
+    reqwest_client: Client,
+    /// The extension id that will be retrieved on register and used
+    /// for subsequent requests
+    extension_id: Option<String>,
+}
+
+impl AwsExtension {
+    /// Creates a new `AwsExtension` instance.
+    pub fn new(aws_runtime_api: String) -> Result<Self, AwsExtensionError> {
+        let base_url = format!("http://{}/2020-01-01/extension", aws_runtime_api);
+
+        let reqwest_client = reqwest::blocking::Client::builder()
+            .timeout(None)
+            .build()
+            .map_err(|_| AwsExtensionError)?;
+
+        Ok(AwsExtension {
+            base_url,
+            reqwest_client,
+            extension_id: None,
+        })
+    }
+
+    fn register(&mut self) -> Result<(), AwsExtensionError> {
+        let url = format!("{}/register", self.base_url);
+        let map = HashMap::from([("events", ["INVOKE", "SHUTDOWN"])]);
+
+        let res = self
+            .reqwest_client
+            .post(&url)
+            .header(EXTENSION_NAME_HEADER, EXTENSION_NAME)
+            .json(&map)
+            .send()
+            .map_err(|_| AwsExtensionError)?;
+
+        if res.status() != StatusCode::OK {
+            return Err(AwsExtensionError);
+        }
+
+        // let register_response: RegisterResponse = res.json().unwrap();
+
+        let extension_id = res
+            .headers()
+            .get(EXTENSION_ID_HEADER)
+            .and_then(|h| h.to_str().ok())
+            .map(|h| h.to_string())
+            .ok_or(AwsExtensionError)?;
+
+        self.extension_id = Some(extension_id);
+        relay_log::info!("AWS extension successfully registered");
+        Ok(())
+    }
+
+    fn next_event(&self) -> Result<NextEventResponse, AwsExtensionError> {
+        match self.extension_id.as_ref() {
+            Some(extension_id) => {
+                let url = format!("{}/event/next", self.base_url);
+
+                self.reqwest_client
+                    .get(&url)
+                    .header(EXTENSION_ID_HEADER, extension_id)
+                    .send()
+                    .and_then(|r| r.json())
+                    .map_err(|_| AwsExtensionError)
+            }
+            None => Err(AwsExtensionError),
+        }
+    }
+}
+
+impl Actor for AwsExtension {
+    type Context = Context<Self>;
+
+    fn started(&mut self, context: &mut Self::Context) {
+        relay_log::info!("AWS extension started");
```

Comment:
Should we talk about "Sentry Lambda Extension" instead of "AWS extension" in log statements?

Reply:
because this code is running in the customers infrastructure and all the output ends up in the customers logs.

Reply:
I'm actually not sure if we want to pollute people's logs at all unless they turn it on explicitly somehow.

Reply:
I will cleanup the logging later, I want it all on for now till we test/benchmark and then we will do one final cleanup